File: Python/Leetcode Medium/133CloneGraph.py
#https://leetcode.com/problems/clone-graph/description/
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        visited = {}
        def helper(node):
            if node == None: return
            if node in visited: return visited[node]
            copy = Node(node.val)
            visited[node] = copy
            for neighbor in node.neighbors:
                copy.neighbors.append(helper(neighbor))
            return copy
        return helper(node)

File: C++/Leetcode Medium/133CloneGraph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
#include <unordered_map>
class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<int, Node*> map;
        return helper(node, map);
    }
    Node* helper(Node* node, unordered_map<int, Node*>& map) {
        if (node == nullptr) return nullptr;
        Node* res = new Node(node->val);
        map[node->val] = res;
        for (Node* neighbhor : node->neighbors) {
            if (map.find(neighbhor->val) != map.end()) res->neighbors.push_back(map[neighbhor->val]);
            else res->neighbors.push_back(helper(neighbhor, map));
        }
        return res;
    }
};